### PR TITLE
Feature/2072 swagger oas

### PR DESCRIPTION
**Pull request by davidnich**

### CODE REVIEW

**Comment by davidnich** [comment]
@omusil24 please do not merge yet

**Comment by davidnich** [comment]
@omusil24 now it should be OK

**Comment by davidnich** [comment]
@omusil24 please do not merge; found some path matching issues related to `basePath`

**Comment by davidnich** [comment]
@omusil24 now it should be OK

File: qlib/Mime.qm [hunk 1]
```diff
@@ -1352,4 +1413,185 @@ hash: (2 members)
...
+        }
+
+        #! adds a message part to the start of the list
+        /** @param data the data for the part
+            @param name the unique name for the part
+            @param hdr an optional hash of headers for the part
+
+            @throw MULTIPART-FORMDATA-ERROR the name parameter is empty or conflicts with an existing part name
+         */
+        splicePart(data data, string name, *hash hdr) {
```
Comment (omusil24) [inline qlib/Mime.qm]: Could you change the `data` variable name so it doesn't collide with the `data` type?

Comment (davidnich) [reply qlib/Mime.qm]: done

File: qlib/Swagger.qm [hunk 2]
```diff
@@ -0,0 +1,3666 @@
...
+    }
+}
+
+#! The object provides metadata about the API. The metadata can be used by the clients if needed, and can be presented in the %Swagger-UI for convenience.
+public class InfoObject inherits ObjectBase {
+    public {
+        #! Required. The title of the application.
+        string title;
+
+        #! A short descripti
```
Comment (omusil24) [inline qlib/Swagger.qm]: There is no `float` swagger type, only `number` and that is equivalent to Qore's `float`... For specifying more precisely, there's the `format` field.

File: qlib/Swagger.qm [hunk 3]
```diff
@@ -0,0 +1,3666 @@
...
+        *int maxItems;
+
+        #! See https://tools.ietf.org/html/draft-fge-json-schema-validation-00#section-5.3.3.
+        *int minItems;
+
+        #! See https://tools.ietf.org/html/draft-fge-json-schema-validation-00#section-5.3.4.
+        *bool uniqueItems;
+
+        #! See https://tools.ietf.org/html/draft-fge-json-schema-validation-00#section-5.5.1.
+        hash<string, bool> enum;
```
Comment (omusil24) [inline qlib/Swagger.qm]: This is wrong because according to documentation, the enum values can have any type, not just `string`.

Comment (omusil24) [reply qlib/Swagger.qm]: And as seen in `getExampleValue()`, not just `int`, `number` or `boolean` (`float` is not a swagger type).

File: qlib/Swagger.qm [hunk 4]
```diff
@@ -0,0 +1,3666 @@
...
+            }
+            case "yaml": {
+%ifdef NoYaml
+                throw "YAML-MODULE-MISSING", "Trying to parse a YAML schema, but the yaml module is unavailable";
+%else
+                rv = parse_yaml(str);
+                break;
+%endif
+            }
+            case "xml": {
```
Comment (omusil24) [inline qlib/Swagger.qm:663]: Why is there an option for XML when Swagger specifications are only JSON or YAML?

File: qlib/Swagger.qm [hunk 5]
```diff
@@ -0,0 +1,3666 @@
...
+        @param opts options as per @ref SwaggerOptions as follows:
+        - \c compact_serialization (default @ref Qore::True "True"): optimize the size of serialized data strings
+        - \c def_path: the default path to use when retrieving external schema references
+        - \c try_import: a call reference or closure to be passed a string name for external schema references, must take a string argument (the resource name) and return a string (the resource data)
+
+        @throws INVALID-FIELD-FORMAT field has invalid format
+        @throws INVALID-FIELD-TYPE field has invalid type
+        @throws REQUIRED-FIELD-MISSING required field is missing
+        @throws SECURITY-ERROR inconsistent security information in schema
+     */
```
Comment (omusil24) [inline qlib/Swagger.qm]: `SWAGGER-OPTION-ERROR`, `INVALID-SWAGGER-VERSION` and `INVALID-FIELD-VALUE` exceptions are missing here.

File: qlib/Swagger.qm (hunk 2 above)
Comment (davidnich) [reply qlib/Swagger.qm]: fixed

File: qlib/Swagger.qm (hunk 3 above)
Comment (davidnich) [reply qlib/Swagger.qm]: updated implementation with example values and implementation notes to document that a hash is used for quick lookups and that we only support enum values that are convertible to a string for practical reasons

File: qlib/Swagger.qm (hunk 4 above)
Comment (davidnich) [reply qlib/Swagger.qm:663]: Swagger 2.0 specifies XML de/serialization, however it has not yet been implemented yet

File: qlib/Swagger.qm (hunk 5 above)
Comment (davidnich) [reply qlib/Swagger.qm]: done

File: qlib/Swagger.qm (hunk 2 above)
Comment (omusil24) [inline qlib/Swagger.qm]: Unused members here.

File: qlib/Swagger.qm [hunk 6]
```diff
@@ -0,0 +1,3683 @@
...
+                    );
+                }
+                if (!MimeDataTypes{content})
+                    continue;
+                return rv += (
+                    "body": compact_serialization ? MimeDataTypes{content}.serialize(response_body) : MimeDataTypes{content}.serialize_verbose(response_body),
+                    "hdr": rv.hdr + ("Content-Type": content),
+                );
+            }
+            if (content_types == MimeContentType
```
Comment (omusil24) [inline qlib/Swagger.qm]: This seems like it is not implemented.

Comment (omusil24) [inline qlib/Swagger.qm]: This seems like it is not implemented.

**Comment by omusil24** [comment]
@davidnich the inline comment adding is broken now for me so I will write further stuff here.

Swagger.qm:1746 - there's a TODO and it seems like it's not implemented
1637-1656 - OperationObject members which seem unused (`operations` hash defined on line 1686 seems to be used instead) - should be removed if that's so
2338, 2341 - also seems unused (since you added inheritance to the parameters)
2544, 2866, 3063 - proper checking of default values for array (of array) and file types is not implemented and untested

**Comment by davidnich** [comment]
@omusil24 
> Swagger.qm:1746 - there's a TODO and it seems like it's not implemented
> 1637-1656 - OperationObject members which seem unused (operations hash defined on line 1686 seems to be used instead) - should be removed if that's so
> 2338, 2341 - also seems unused (since you added inheritance to the parameters)
> 2544, 2866, 3063 - proper checking of default values for array (of array) and file types is not implemented and untested

All issues have been addressed

**Pull request merged**